File: 198-house-robber/198-house-robber.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        n = len(nums)-1
        if n==0: return nums[n]
        dp = [-1]*(n+1)
        def solve(ind,dp):
            if ind==0: return nums[ind]
            if ind<0: return 0
            
            if dp[ind]!=-1: return dp[ind]
            
            pick = nums[ind] + solve(ind-2,dp)
            not_pick = solve(ind-1,dp)
        
            dp[ind] = max(pick,not_pick)
            return dp[ind]
        
        solve(n,dp)
        return dp[n]

File: 198-house-robber/198-house-robber.cpp
class Solution {
public:
    int solve(int ind,vector<int>& nums,vector<int>& dp) {
        if(ind<0) return 0;
        
        if(dp[ind]!=-1) return dp[ind];
        return dp[ind] = max(nums[ind] + solve(ind-2,nums,dp),solve(ind-1,nums,dp));
    }
    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n,-1);
        return solve(n-1,nums,dp);
    }
};